kratos/geometries/integration_point_surface_3d.h

```diff
@@ -0,0 +1,473 @@
+     * gradients. ShapefunctionsLocalGradients function return this
+     * type as its result.
+     */
+    typedef typename BaseType::ShapeFunctionsGradientsType ShapeFunctionsGradientsType;
+
+    /**
+     * A third order tensor to hold shape functions' local second derivatives.
+     * ShapefunctionsLocalGradients function return this
+     * type as its result.
+     */
+    typedef typename BaseType::ShapeFunctionsSecondDerivativesType
+        ShapeFunctionsSecondDerivativesType;
+
+
+    /**
+    * A fourth order tensor to hold shape functions' local third derivatives.
+    * ShapefunctionsLocalGradients function return this
+    * type as its result.
+     */
+    typedef typename BaseType::ShapeFunctionsThirdDerivativesType
+        ShapeFunctionsThirdDerivativesType;
+
+    /**
+    * Type of the normal vector used for normal to edges in geomety.
+     */
+    typedef typename BaseType::NormalType NormalType;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    IntegrationPointSurface3d(
+        const PointsArrayType& ThisPoints,
+        const CoordinatesArrayType& rLocalCoordinates,
+        const double& rIntegrationWeight,
+        Matrix& rShapeFunctions)
+        : BaseType( ThisPoints, nullptr )
+    {
+        IntegrationPoint(rLocalCoordinates[0], rLocalCoordinates[1], rIntegrationWeight);
+
+        IntegrationPoints = IntegrationPointsContainerType(1);
+        IntegrationPoints[0] = IntegrationPoint;
+
+        msGeometryData.IntegrationPoints(IntegrationPoints);
+
+        KRATOS_ERROR_IF(rShapeFunctions.size2() != ThisPoints.size()) <<
+            "Length of shape function array in matrix does not fit to number of points. Length of point vector: " << ThisPoints.size()
+            << " Length of corresponding shape function vector: " << rShapeFunctions.size2() << std::endl;
+
+        // shape functions
+        if (rShapeFunctions.size1() > 0)
+        {
+            Matrix ShapeFunctionsValues = ZeroMatrix(1, PointsNumber());
+            for (IndexType i = 0; i < PointsNumber(); ++i)
+            {
+                ShapeFunctionsValues(0, i) = rShapeFunctions(0, i);
+            }
+            msGeometryData.ShapeFunctionsValues(ShapeFunctionsValues);
+        }
+
+        // first derivatives/ gradients
+        if (rShapeFunctions.size1() > 2)
+        {
+            ShapeFunctionsGradientsType ShapeFunctionsLocalGradients(1);
+            Matrix ShapeFunctionsLocalGradientsIntegrationPoint = ZeroMatrix(2, PointsNumber());
+            for (IndexType i = 0; i < PointsNumber(); ++i)
+            {
+                ShapeFunctionsLocalGradientsIntegrationPoint(i, 0) = rShapeFunctions(1, i);
+                ShapeFunctionsLocalGradientsIntegrationPoint(i, 1) = rShapeFunctions(2, i);
+            }
+            ShapeFunctionsLocalGradients.push_back(ShapeFunctionsLocalGradientsIntegrationPoint);
+            msGeometryData.ShapeFunctionsLocalGradients(ShapeFunctionsLocalGradients);
```

Comment:
Are you sure that you want to change the Geometry Data for all objects of type `IntegrationPointSurface3d`? 

If the number of integration points depends to the number of points you cannot use the geometrical data and we should make the correspon methods virtual

Reply:
Actually, I don't see a blocker and for all other cases it is more efficient.

I see GeometryData only as data container. E.g. in Iga we don't follow standard integration rules (especially for trimmed geometries). Thus, I want a possibility to fill up the container with data which is computed somewhere but not necessariliy at construction time and not from inside the geometry (The reason is it should be independent on shape function. NURBS/ linear FEM/ whatever...). Those setters are the only functions with which I added to the GeometryData. With those, geometry.h does not has to be touched.

Anyways, please see this IntegrationPointGeometry as something special. It shall be the latest entitity in the chain. Due to this generity it can be used anywhere. This is nice for e.g. MPM (change the background whenever to whatever - the shape function type is defined), in IGA many different ways to discretize surface have their validity and are done from outside and latest this was not discussed fully yet, but for interfaces between physics this can be also used, with this we would have a standardized form anywhere.

Maybe this figure describes it a bit better:
![image](https://user-images.githubusercontent.com/25983517/56802579-b23b8c80-6820-11e9-8f56-334d2ca7e633.png)


Later the BrepGeometries might not even need this modification in the GeometryData. But, let's see this.

Reply:
My question are all surfaces share the same gradient array in entire Kratos?

Reply:
Actually, first, I didn't understand your issue. Now I do. And you're right, we cannot handle this with this static GeometryData member.

Gradients and Shape Functions are different for each of those geometries. But, we can keep at least dimensions, ...